### ISSUE TITLE
Pretrained model performance lower than expected

### ISSUE THREAD

**Issue by kaimingkuang** [salesforce/ULIP#6]
Hi, I am following the README instructions (downloaded the data and set the soft links) to run the pretraining task. This is how I run pretraining with the PointNet++ backbone:
```bash
python main.py --output-dir=./outputs/ulip_pointnet2_ssg --model=ULIP_PN_SSG
```
I also run the pretraining with PointMLP, with [the model code copied from the PointMLP repo](https://github.com/ma-xu/pointMLP-pytorch/blob/main/classification_ModelNet40/models/pointmlp.py#L349). Here is how I modified the `ULIP_models.py` file to include PointMLP backbone:
```python
def ulip_pointmlp(args):
    vision_model = timm.create_model('vit_base_patch16_224', num_classes=0)

    # =====================================================================
    # This is a sample template to pre-train your customized 3D backbones, please modify this part accordingly!
    pc_feat_dims = 512
    point_encoder = PointMLP(class_num=pc_feat_dims, embed_dim=64, groups=1, res_expansion=1.0,
        activation="relu", bias=False, use_xyz=False, normalize="anchor",
        dim_expansion=[2, 2, 2, 2], pre_blocks=[2, 2, 2, 2], pos_blocks=[2, 2, 2, 2],
        k_neighbors=[24, 24, 24, 24], reducers=[2, 2, 2, 2],
        in_channel=3, points=1024)
    # We assume you might have different point cloud output feature dimension,
    # we added a projecting layer to unify the point cloud output dimension before doing the multimodal alignment,
    # please change the output feature dimension here.
    # =====================================================================

    model = ULIP_WITH_IMAGE(embed_dim=512, vision_width=768, point_encoder=point_encoder, vision_model=vision_model,
                            context_length=77, vocab_size=49408,
                            transformer_width=512, transformer_heads=8, transformer_layers=12, pc_feat_dims=pc_feat_dims)

    if not args.evaluate_3d:
        # load the pretrained model
        pretrain_slip_model = torch.load('./data/initialize_models/slip_base_100ep.pt', map_location=torch.device('cpu'))
        pretrain_slip_model_params = pretrain_slip_model['state_dict']
        pretrain_slip_model_params = {param_name.replace('module.', ''): param for param_name, param in
                                      pretrain_slip_model_params.items()}

        for name, param in model.named_parameters():
            if name not in pretrain_slip_model_params:
                continue

            if isinstance(pretrain_slip_model_params[name], Parameter):
                param_new = pretrain_slip_model_params[name].data
            else:
                param_new = pretrain_slip_model_params[name]

            param.requires_grad = False
            print('load {} and freeze'.format(name))
            param.data.copy_(param_new)

    return model
```
Then I run:
```bash
python main.py --output-dir=./outputs/ulip_pointmlp --model=ulip_pointmlp
```
However, the model performances (zero-shot on ModelNet40) are lower than expected. PointMLP OA even droped to a extremely low point:
![W B Chart 3_9_2023, 8_00_27 PM](https://user-images.githubusercontent.com/62201355/224220137-09a1287c-b1a4-4f5d-b761-9442c859ab09.png)
Seems like the PointNet++ has already converged and it only reached 50% OA. Am I missing something here? Thank you!

**Comment by Tycho-Xue** [comment]
@kaimingkuang Hi, I just reproduced the pointnet++ ssg again for release purposes, used the exact same codebase, and I have uploaded the checkpoint, please check that out.
I think by default it's pre-trained with 250 epochs, with 8 A100 it should take around 13 hrs to finish using the default setting.
For PointMLP, I'm trying to see the difference from my pointmlp code: instead of set class_num=pc_feat_dims, try to remove the classifier's last layer, and set the pc_feat_dims = 256. I think it's a more natural way to peal off the last layer and then pre-train the rest.

self.classifier = nn.Sequential(
            nn.Linear(last_channel, 512),
            nn.BatchNorm1d(512),
            self.act,
            nn.Dropout(0.5),
            nn.Linear(512, 256),
            nn.BatchNorm1d(256),
            self.act,
            nn.Dropout(0.5),
        )

**Comment by Tycho-Xue** [comment]
@kaimingkuang this is the test acc1 graph for my reproduced pre-training.
![image](https://user-images.githubusercontent.com/31913279/224235817-729213fd-0be1-45a8-b235-156c7e9979dd.png)

**Comment by Colin97** [comment]
> @kaimingkuang Hi, I just reproduced the pointnet++ ssg again for release purposes, used the exact same codebase, and I have uploaded the checkpoint, please check that out. I think by default it's pre-trained with 250 epochs, with 8 A100 it should take around 13 hrs to finish using the default setting. For PointMLP, I'm trying to see the difference from my pointmlp code: instead of set class_num=pc_feat_dims, try to remove the classifier's last layer, and set the pc_feat_dims = 256. I think it's a more natural way to peal off the last layer and then pre-train the rest.
> 
> self.classifier = nn.Sequential( nn.Linear(last_channel, 512), nn.BatchNorm1d(512), self.act, nn.Dropout(0.5), nn.Linear(512, 256), nn.BatchNorm1d(256), self.act, nn.Dropout(0.5), )

Do you mean pc_feat_dims=512? Since SLIP feat dim is 512

**Comment by Tycho-Xue** [comment]
By default pc_dims is your backbone’s output feature dimension, here your
output dimension is 256

**Comment by Tycho-Xue** [comment]
@kaimingkuang @Colin97 I think I figured out your pointMLP problem, since you are using the bash scripts from the pointnet2, the default lr is 3e-3 for pointnet2, you can change it to 1e-3 for pointMLP, somehow pointMLP is more sensitive to the lr, I just reproduced pointMLP results using he released codebase, it works fine. Let me know if it still doesn't work this time, thanks

**Comment by Tycho-Xue** [comment]
@kaimingkuang @Colin97 FYI, more supported backbones and reproduced checkpoints are released

**Comment by auniquesun** [comment]
> 
@Tycho-Xue  Thanks for sharing the paper and code. Novel and excellent work !

I am very interested this work and I read the paper and code. 
One point I wonder is why the pre-training cost is so **expensive** (**8 A100 GPUs, 13 hours training on ShapeNet**), since you have already frozen the weights in image and text branches? 

According to my experiences, the point cloud encoders are much more lightweight than the image/text counterparts thus the training of point cloud encoders would be much more efficient. In my opinion, a research team that has 2~4 V100 GPUs can follow your work but according to your report, using 8 A100 GPUs is a surprise for me. 

I think more than 90% research groups can not afford for the cost of 8 A100 GPUs.

**Comment by Tycho-Xue** [comment]
Hi @auniquesun, thanks for your concern, we used 8 A100 GPUs for faster experiments iterations and also it's a common machine node type we use when doing pre-training. However, although we didn't tweak around the pre-training machine settings, we think it's workable to pre-train it on a smaller scale of machines which most research groups should be able to provide, we just didn't tweak on the machine settings before, feel free to tweak around it.

**Comment by LiXinghui-666** [comment]
> 

Hi @Tycho-Xue, I would like to ask if I want to retrain the point cloud encoder, can it be trained normally **without the 8 A100 GPUs**? Since our team **only has 8 A10 GPUs**(24G memory).

**Comment by Tycho-Xue** [comment]
@LiXinghui-666 , I think it should work, it's try to pre-train with 8A10 GPUs with the default batch size (64) first, if it has OOM, then lower the batch size until it works, I have tried with lower batch size, and I found ULIP's framework is robust to it and still performs good.
also a smaller backbone will need smaller mem, for example, pointbert need more mem than backbones like pointnext.
We used 8 A100 for faster iteration, not meaning it's a minimum requiremen to pre-train ULIP.
Hope it helps

**Comment by StanLei52** [comment]
> @kaimingkuang this is the test acc1 graph for my reproduced pre-training. ![image](https://user-images.githubusercontent.com/31913279/224235817-729213fd-0be1-45a8-b235-156c7e9979dd.png)

Hi @Tycho-Xue, it seems that Acc@1 for zero-shot classification on ModelNet does not achieve 60.3% in this pic?

**Comment by Tycho-Xue** [comment]
Hi, this screenshot is for pointnet_ssg backbone as indicated on the top
right corner, 60.3 is for pointbert backbone.
